### PR TITLE
Flexible output sizes

### PR DESCRIPTION
**Pull request by ericmjl**
In this PR, we provide the ability to specify one of Unirep-64, Unirep-256, or Unirep-1900 outputs, with the goal of making the code architecture eventually flexible enough to handle arbitrarily-specified mLSTM sizes (i.e. 64/256/1900/anything).

I'm working out a better API here for the `fit` function, in particular one that allows us to evotune from random starter weights. This I think might be the most common use case a protein engineer will find themselves in.

When we switch from an mLSTM1900-default API to an mLSTM-any-size API, I think the key parameter we need to expose in the relevant functions is `mlstm_size`, which would be 1900 for our default models, and any integer otherwise.

@ElArkk and @ivanjayapurna if you both have some time, could you take a look at it and see whether you have any feedback on the API design? The change is an API-breaking change, so I don't want to rush this one.

### CODE REVIEW

**Comment by ElArkk** [comment]
Really cool stuff @ericmjl , one step closer to a fully plug-and-play RNN model. I have just pushed some changes to make it easier for non-expert users to still use `fit` and `evotune` with the paper weights. Let me know what you think of having default values for `mlstm_size` and `rng`, especially for `rng` I'm not quite sure if it's good to have a default value.

One other thing: In the original paper, the 256 and 64 unit mLSTM actually consist of 4 stacked layers of those mLSTMs. I think we are in a good position to allow for this quite easily using `stax`, but it would need some more changes of how we pass a model and params to `fit`. Maybe this is for a seperate PR.

File: jax_unirep/evotuning.py [hunk 1]
```diff
@@ -249,7 +251,10 @@ def step(i, x, y, state):
 
     # Load and check that params have correct keys and shapes
     if params is None:
-        _, params = init_func(rng=rng, input_shape=(-1, 10))
+        if mlstm_size == 1900:
+            params = load_params()
+        else:
+            _, params = init_func(rng=rng, input_shape=(-1, 10))
```
Comment (ericmjl) [inline jax_unirep/evotuning.py]: Minor nitpick, but I think the original is a bit more concise, and accomplishes the same thing. (Initialize with something, clobber over if another condition is met.)

File: jax_unirep/evotuning.py [hunk 2]
```diff
@@ -485,20 +490,27 @@ def evotune(
...
     but if you want to use randomly intialized weights:
 
-        from jax_unirep.evotuning import evotuning_funcs
+        from jax_unirep.evotuning import evotuning_funcs, fit
         from jax.random import PRNGKey
 
-        init_func, _ = evotuning_funcs(mlstm_size=256)
+        init_func, _ = evotuning_funcs(mlstm_size=256) # works for any size
         _, params = init_func(PRNGKey(0), input_shape=(-1, 10))
+        fit_func = partial(fit, mlstm_size=256, params=params)
```
Comment (ericmjl) [inline jax_unirep/evotuning.py]: Danke!!!! I missed out on updating the docs. Noice catch.

File: jax_unirep/evotuning.py [hunk 3]
```diff
@@ -485,20 +490,27 @@ def evotune(
...
 
     By default, mLSTM and Dense weights from the paper are used
-    by passing in `params=None`,
+    by setting `mlstm_size=1900` and `params=None`
+    in the partially-evaluated fit function (`fit_func`),
     but if you want to use randomly intialized weights:
 
-        from jax_unirep.evotuning import evotuning_funcs
+        from jax_unirep.evotuning import evotuning_funcs, fit
         from jax.random import PRNGKey
```
Comment (ericmjl) [inline jax_unirep/evotuning.py:498]:
```suggestion
        from jax_unirep.evotuning import evotuning_funcs, fit
        from jax.random import PRNGKey
        from functools import partial
```

File: jax_unirep/evotuning.py (hunk 1 above)
Comment (ElArkk) [reply jax_unirep/evotuning.py]: What I wanted to accomplish here is defaulting to the paper weights if someone doesn't specify `params` for an `mlstm_size` of 1900. But I'm also not 100% happy with this solution. Maybe instead of giving a lot of defaults, we just need to make it very clear in the documentation of how one can evotune starting with the weights from the paper.

Comment (ericmjl) [reply jax_unirep/evotuning.py]: Hmmm yeah. I think I'm in favour of putting in the documentation how to load weights from the paper. The added complexity can creep in over time, leading to code duplication (that I tried my best in this PR to remove as well). 

When combined with the weights licensing issue, I can see how some teams (hint hint) might want to avoid using the pre-trained weights. From prior experience with things like hierarchical and local models, it seems to me that making it easy to train "local" weights is going to be a more pragmatically productive thing than training on entire large datasets.

Comment (ElArkk) [reply jax_unirep/evotuning.py]: agreed!

**Comment by ericmjl** [comment]
@ElArkk I think I finally get how to design this API in an easy fashion.

Pardon me, this comment might be long. I'm also going to record our conclusions in the docs once we're done debating it.

I think I finally see where you were going with the "if output size is 1900, then give us the 1900-weights". That said, there's a bit too much magic in there, IMO, as the mLSTM output size isn't sufficient to specify the full architecture. While yes, the 1900-model only has one layer, while the 256-model has four layers, there's no _a priori_ reason to restrict the 256-model to four layers. We could just as well do three, or five. In other words, the mapping from the integer "mLSTM output dim" to the integer "number of layers" isn't _necessarily_ a 1-to-1 mapping.

I think the better way to handle this situation is as follows.

Firstly, we add the `model_func` and `params` kwargs to the `fit` function.

Secondly, we set the `model_func` and `params` kwargs to the default mLSTM1900 model function and params. If a user wants to deviate from the mLSTM1900 function, then they must provide the appropriate model function and params on their own. `model_func` is a function that accepts only `params` and `x`. Basically the `apply_fun` that gets returned from `stax.serial`.

So how do we conveniently allow users to deviate from the mLSTM1900 model, and say, evotune the mLSTM256 or mLSTM64 models? We can provide those as functions too :).

I think the flexibility issue was the piece that was troubling me the most on this PR. "How do we provide flexibility _and_ convenience at the same time?" Turns out, there's always some design pattern available to enable that, I think, with different levels of compromise.

**Comment by ElArkk** [comment]
@ericmjl Yes, my initial solution definitely has too much magic. Now with the nicely refactored `evotuning layers` and `evotuning_funcs` functions I guess it's trivial to allow for arbitrary model architectures. We also already have checks in place that the number of layers of the specified model match with the number of parameters passed.  We would probably need to rethink how we applied the `validate_mLSTM_params` function as a defensive check in fit though, as we can't be sure about the ordering of layers anymore.

About the `model_func`: You propose that we  bypass lines 274 and 275 in `fit` when the user specifies a model, right? I think that's a good idea, because this way, we can more easily allow for any kind of architecture (even parallel ones I guess?). And we can have convenience functions to load the right `model_funcs` for the "standard" architectures 1900, 256 and 64.

So then, we have three levels of control:

- The user uses all defaults, we give them the mLSTM1900 with paper weights.
- The user specifies a custom `model_func`, but no `params`. We randomly initialise `params` for them.
- The user specifies both a custom `model_func` and `params`.

As long as we have a clear example in the docs about how to accomplish all this I think it's pretty straight-forward to use.

**Comment by ericmjl** [comment]
> So then, we have three levels of control:
> 
> - The user uses all defaults, we give them the mLSTM1900 with paper weights.
> - The user specifies a custom model_func, but no params. We randomly initialise params for them.
> - The user specifies both a custom model_func and params.

Yes, the clarity of thought there! Great, this comment helps me with writing out what the docs should look like. Danke sir!

**Comment by ericmjl** [comment]
@ElArkk writing this down here so I can gain some clarity of thought.

Parameter validation as it stands right now is tied to a strong assumption of "model == mLSTM1900". Model parameters, however, should be done slightly differently, I think. Model params and their shapes depend highly on the architecture that is specified. Thus, I think parameter validation should simply be that dummy data are passed through the model with no errors:

```python
from jax.random import normal, PRNGKey
from jax import vmap
from functools import partial
import warnings

def validate_params(model_func, params):
    dummy_embedding = normal(PRNGKey(42), shape=(2, 3, 10)  # n_samps, n_letters, n_embed_dims
    try:
        vmap(partial(model_func, params))(dummy_embedding)
    except Exception as e:
        warnings.warn("You may have shape issues! Check that your params are of the correct shapes for the specified model. Here's the original warning below:")
        raise e
```

File: jax_unirep/evotuning.py [hunk 4]
```diff
@@ -98,23 +99,75 @@ def batch_iter(
...
+
 def fit(
-    params: Dict,
-    sequences: Set[str],
+    sequences: Iterable[str],
     n_epochs: int,
-    batch_method: Optional[str] = "random",
-    batch_size: Optional[int] = 25,
+    model_func: Callable = mlstm1900_apply_fun,
+    params: Any = load_params(),
```
Comment (ElArkk) [inline jax_unirep/evotuning.py:163]: Would it be better to set default to `None` and conditionally call on `load_params` below, such that users can still explicitly set `params=None`. I think I'd be better for code readability.

Comment (ericmjl) [reply jax_unirep/evotuning.py:156]: I think I see where you're coming from with this one. Makes sense, the pattern you're suggesting:

```python
fitted_params = fit(..., params=None,...)
```

makes it possible for an end-user to _explicitly_ indicate that no params have been passed in. OK, let me change.

**Comment by ericmjl** [comment]
@ElArkk I think I'm ready for final review on this one!

File: jax_unirep/evotuning.py [hunk 5]
```diff
@@ -82,39 +81,86 @@ def avg_loss(
...
-            sum_loss += evotune_loss_jit(params, inputs=x, targets=y) * len(x)
+            sum_loss += evotune_loss_jit(
+                params=params, inputs=x, targets=y
+            ) * len(x)
             num_seqs += len(x)
 
     return sum_loss / num_seqs
 
 
+def generate_one_length_batch(
```
Comment (ElArkk) [inline jax_unirep/evotuning.py:103]:
```suggestion
def generate_single_length_batch(
```

Maybe this describes the purpose of the function a bit better?

**Comment by ericmjl** [comment]
@ElArkk, I think we're ready to merge! At least I think after merging, I'm game to try out some new ideas, such as dataset distillation or something like that, using UniRep.

**Comment by ElArkk** [comment]
Hitting the button! And yes, definitely many cool things to try out still :)
